python/mspasspy/global_history/manager.py

```diff
@@ -0,0 +1,261 @@
+
+        if not alg_name:
+            # get the algorithm name
+            alg_name = alg[:alg.index('(')].strip()
+        if not parameters:
+            # get the parameters
+            parameters = alg[alg.index('(')+1:].strip()
+        
+        # get the alg_id if exists, else create a new one
+        alg_id = global_history.get_alg_id(alg_name, parameters)
+        if not alg_id:
+            alg_id = ObjectId()
+        global_history.logging(alg_name, alg_id, parameters)
+    return self.reduce(func, *args, **kwargs)
+
+def mspass_dask_reduce(self, func, *args, global_history=None, alg_name=None, parameters=None, **kwargs):
+    """
+     This decorator method add more functionaliy on the standard dask reduce method and be a part of member functions
+     in the dask bag library. Instead of performing the normal reduce function, if user provides global history manager,
+     alg_name and parameters as input, the global history manager will log down the usage of the algorithm.
+
+    :param func: target function
+    :param global_history: a user specified global history manager
+    :type global_history: :class:`GlobalHistoryManager`
+    :param alg_name: a user specified alg_name for the reduce operation
+    :type alg_name: :class:`str`
+    :param parameters: a user specified parameters for the reduce operation
+    :type parameters: :class:`str`
+    :return: a dask `bag` format of objects.
+    """
+    if global_history:
+        # get the whole lambda function
+        alg_string = getsource(func)
+        # truncate to get the algorithm name and parameters
+        alg = alg_string[alg_string.index(':')+1 : alg_string.index(')')]
+
+        if not alg_name:
+            # get the algorithm name
+            alg_name = alg[:alg.index('(')].strip()
+        if not parameters:
+            # get the parameters
+            parameters = alg[alg.index('(')+1:].strip()
+
+        # get the alg_id if exists, else create a new one
+        alg_id = global_history.get_alg_id(alg_name, parameters)
+        if not alg_id:
+            alg_id = ObjectId()
+        global_history.logging(alg_name, alg_id, parameters)
+    return self.fold(func, *args, **kwargs)
+
+class GlobalHistoryManager:
+    """
+    A Global History Mananger handler.
+
+    This is a handler used in the mspass_client, normally user should not directly create
+    a Global History Manager by his own. Instead, user should get the Global History Manager
+    through mspass client's methods.
+    """
+    def __init__(self, database_instance, job_name, collection=None):
+        self.job_name = job_name
+        # generate an bson UUID for this job, should be unique on the application level
+        self.job_id = ObjectId()
+
+        self.history_db = database_instance
+
+        self.collection = collection
+        if not self.collection:
+            # use the `history` collection defined in database schema
+            self.collection = 'history'
```

Comment:
Need to update mspass.yaml and use the `default_name` method to get the default collection name.